packages/program/src/contract-call-script.ts

```diff
@@ -91,55 +120,112 @@ const getMainCallReceipt = (
       type === ReceiptType.Call && from === SCRIPT_WRAPPER_CONTRACT_ID && to === contractId
   );
 
-const scriptResultDecoder = (contractId: AbstractAddress) => (result: ScriptResult) => {
-  if (toNumber(result.code) !== 0) {
-    throw new Error(`Script returned non-zero result: ${result.code}`);
-  }
+const scriptResultDecoder =
+  (contractId: AbstractAddress, isOutputDataHeap: boolean) => (result: ScriptResult) => {
+    if (toNumber(result.code) !== 0) {
+      throw new Error(`Script returned non-zero result: ${result.code}`);
+    }
 
-  const mainCallResult = getMainCallReceipt(
-    result.receipts as TransactionResultCallReceipt[],
-    contractId.toB256()
-  );
-  const mainCallInstructionStart = bn(mainCallResult?.is);
-  const receipts = result.receipts as ReturnReceipt[];
-  return receipts
-    .filter(({ type, is }) => isReturnType(type) && mainCallInstructionStart.eq(bn(is)))
-    .map((receipt: ReturnReceipt) => {
-      if (receipt.type === ReceiptType.Return) {
-        return new U64Coder().encode((receipt as TransactionResultReturnReceipt).val);
-      }
-      if (receipt.type === ReceiptType.ReturnData) {
-        const encodedScriptReturn = arrayify(receipt.data);
-        return encodedScriptReturn;
-      }
+    const mainCallResult = getMainCallReceipt(
+      result.receipts as TransactionResultCallReceipt[],
+      contractId.toB256()
+    );
+    const mainCallInstructionStart = bn(mainCallResult?.is);
 
-      return new Uint8Array();
-    });
-};
+    const receipts = result.receipts as ReturnReceipt[];
+    return receipts
+      .filter(({ type }) => isReturnType(type))
+      .flatMap((receipt: ReturnReceipt, index, filtered) => {
+        if (!mainCallInstructionStart.eq(bn(receipt.is))) {
+          return [];
+        }
+        if (receipt.type === ReceiptType.Return) {
+          return [new U64Coder().encode((receipt as TransactionResultReturnReceipt).val)];
+        }
+        if (receipt.type === ReceiptType.ReturnData) {
+          const encodedScriptReturn = arrayify(receipt.data);
+          if (isOutputDataHeap && isReturnType(filtered[index + 1]?.type)) {
+            const nextReturnData: TransactionResultReturnDataReceipt = filtered[
+              index + 1
+            ] as TransactionResultReturnDataReceipt;
+            return concat([encodedScriptReturn, arrayify(nextReturnData.data)]);
+          }
+
+          return [encodedScriptReturn];
+        }
+
+        return [new Uint8Array()];
+      })
+      .filter((v) => v !== undefined);
```

Comment:
All of the `return` statements in `flatMap` return an array. How is it possible for this to ever be `undefined` so that filtering has to be done for it?